code/datums/components/surgery_initiator.dm

```diff
@@ -0,0 +1,340 @@
+	var/mob/living/last_user = last_user_ref?.resolve()
+	if (!isnull(last_user_ref))
+		UnregisterSignal(last_user, COMSIG_MOB_SELECTED_ZONE_SET)
+
+	var/mob/living/surgery_target = surgery_target_ref?.resolve()
+	if (!isnull(surgery_target_ref))
+		UnregisterSignal(surgery_target, COMSIG_MOB_SURGERY_STARTED)
+
+/// Does the surgery initiation.
+/datum/component/surgery_initiator/proc/initiate_surgery_moment(datum/source, atom/target, mob/user)
+	SIGNAL_HANDLER
+	if(!isliving(target))
+		return
+	INVOKE_ASYNC(src, PROC_REF(do_initiate_surgery_moment), target, user)
+	return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/datum/component/surgery_initiator/proc/do_initiate_surgery_moment(mob/living/target, mob/user)
+	var/datum/surgery/current_surgery
+
+	for(var/i_one in target.surgeries)
+		var/datum/surgery/surgeryloop = i_one
+		if(surgeryloop.location == user.zone_selected)
+			current_surgery = surgeryloop
+			break
+
+	if (!isnull(current_surgery) && !current_surgery.step_in_progress)
+		attempt_cancel_surgery(current_surgery, target, user)
+		return
+
+	var/list/available_surgeries = get_available_surgeries(user, target)
+
+	if(!length(available_surgeries))
+		if (!(target.mobility_flags & MOBILITY_STAND))
+			target.balloon_alert(user, "no surgeries available!")
+		else
+			target.balloon_alert(user, "make them lie down!")
+
+		return
+
+	unregister_signals()
+
+	last_user_ref = WEAKREF(user)
+	surgery_target_ref = WEAKREF(target)
+
+	RegisterSignal(user, COMSIG_MOB_SELECTED_ZONE_SET, PROC_REF(on_set_selected_zone))
+	RegisterSignal(target, COMSIG_MOB_SURGERY_STARTED, PROC_REF(on_mob_surgery_started))
+
+	ui_interact(user)
+
+/datum/component/surgery_initiator/proc/get_available_surgeries(mob/user, mob/living/target)
+	var/list/available_surgeries = list()
+
+	var/mob/living/carbon/carbon_target
+	var/obj/item/bodypart/affecting
+	if (iscarbon(target))
+		carbon_target = target
+		affecting = carbon_target.get_bodypart(check_zone(user.zone_selected))
+
+	for(var/datum/surgery/surgery as anything in GLOB.surgeries_list)
+		if(!surgery.possible_locs.Find(user.zone_selected))
+			continue
+		if(affecting)
+			if(!surgery.requires_bodypart)
+				continue
+			if(surgery.requires_bodypart_type && !(affecting.bodytype & surgery.requires_bodypart_type))
+				continue
+			if(surgery.requires_real_bodypart && affecting.is_pseudopart)
+				continue
+		else if(carbon_target && surgery.requires_bodypart) //mob with no limb in surgery zone when we need a limb
+			continue
+		if(surgery.lying_required && (target.mobility_flags & MOBILITY_STAND))
+			continue
+		if(!surgery.can_start(user, target))
+			continue
+		for(var/path in surgery.target_mobtypes)
+			if(istype(target, path))
+				available_surgeries += surgery
+				break
+
+	return available_surgeries
+
+/// Does the surgery de-initiation.
+/datum/component/surgery_initiator/proc/attempt_cancel_surgery(datum/surgery/the_surgery, mob/living/patient, mob/user)
+	var/selected_zone = user.zone_selected
```

Comment:
Need to convert all usages of zone_selected to either the setter or combat_zone.

Kinda confusing though.